src/kcas_data/promise.ml

```diff
@@ -0,0 +1,57 @@
+open Kcas
+
+type 'a internal = 'a Magic_option.t Loc.t
```

Comment:
The `Magic_option` avoids a level of indirection.  I decided to go with this optimization to reduce space usage and because promise is already using some other OCaml magic (as does the Eio `Promise` implementation).